src/com/cft/contactmerge/StreetAddressMatchLogic.java

```diff
@@ -2,13 +2,185 @@
+    public static Hashtable<String,String> getRoadTypes(){
+        return roadTypes;
+    }
+
+    public static Hashtable<String,String> getCompassDirection(){
+        return compassDirection;
+    }
+
+    public static ArrayList<String> getPoBoxTypes(){
+        return poBoxTypes;
+    }
+
     /*******************************************************************************************************************
      ************************************************* Matching Methods ************************************************
      *******************************************************************************************************************/
-    public static AnswerType getStreetAddressMatchResult(ArrayList<String> stAddress1, ArrayList<String> stAddress2){
-        // ToDo: Fill in code to obtain matching result
+    public static AnswerType getStreetAddressMatchResult(ArrayList<String> stAddressParts1, ArrayList<String> stAddressParts2){
+        // Precondition: All addressParts must follow the rules from the setNormalizeStreetAddress()
+        String st1Direction = "";
+        String st2Direction = "";
+        String st1RoadType = "";
+        String st2RoadType = "";
+
+        // get return on po box addressses (if either one is a po box)
+        if (isAddressPoBox(stAddressParts1) && isAddressPoBox(stAddressParts2)){
+            // Strip all characters out of address1, and only leave numbers
+            for (int part = 0; part < stAddressParts1.size(); part++){
+                if (!isInteger(stAddressParts1.get(part))){
+                    stAddressParts1.remove(part);
+                    part--;
+                }
+            }
+            // Strip all characters out of address2, and only leave numbers
+            for (int part = 0; part < stAddressParts2.size(); part++) {
+                if (!isInteger(stAddressParts2.get(part))) {
+                    stAddressParts2.remove(part);
+                    part--;
+                }
+            }
+            // if both box addresses match, return yes
+            if (stAddressParts1.equals(stAddressParts2)){
+                return AnswerType.yes;
+            } else {
+                return AnswerType.no;
+            }
+        } else if (isAddressPoBox(stAddressParts1) ^ isAddressPoBox(stAddressParts2)){
+            return AnswerType.no;
+        }
+
+
+        // abstract street direction if the address has one
+        int stDirectionElement = 0;
+        for (String stPart: stAddressParts1){
+            if (getCompassDirection().containsKey(stPart)){
+                st1Direction = getCompassDirection().get(stPart);
+                stAddressParts1.remove(stDirectionElement); // remove the direction from the address
+                break;
+            }
+            stDirectionElement++;
+        }
+        stDirectionElement = 0;
+
+        for (String stPart: stAddressParts2){
```

Comment:
Also duplicate code from above? Split into helper method.

Reply:
This one may be a bit more tricky since it's doing two things. If I seperated this into functions, I would have to write two methods for each address (4 for loops) just for the direction.